### DISCUSSION TITLE
Get NAT64 external route after joining is intermittent

### DISCUSSION THREAD

**Discussion by motters** [openthread/openthread#9667] [Q&A] [answered]
Hi,

When our devices (Zephyr based) boot-up they first join the network and then request the NAT64 external route.

However we're finding that the retrieval of the route is hit and miss. Sometimes obtaining it and other times returning OT_ERROR_NOT_FOUND. 

We attempt retrieving the nat64 route recursively with a 30 second delay between each attempt. However no matter how long we wait the device never gets the route. However sometimes it'll get the route first time.

We have four devices in our mesh for testing (virtually next to each other). Some of the devices will get the nat64 route and some of them will not.

My questions are

- How / when does the device get the external routes?
- Is there a function for the device to request them manually?
- Do you see any glaring issues with the function below?
- Could this be an issue with the device joining via a leader and not the actual border router (ESP32 based)?

Nat64 Code
```cpp
int openthread_controller::find_nat64_prefix()
{
    otNetworkDataIterator iterator = OT_NETWORK_DATA_ITERATOR_INIT;
    otExternalRouteConfig config;
        
    // Get the default instant as we only use one
    otInstance* ot_instance = openthread_get_default_context()->instance;
    
    // Loop through all routes to find a nat64 prefix
    auto result = otNetDataGetNextRoute(ot_instance, &iterator, &config);
    while(result == OT_ERROR_NONE)
    {
	    // If the current route is a nat64 prefix
	    if(config.mNat64)
	    {
		    // Mark that a nat prefix was found
		    m_found_nat64 = true;
    
		    char string[OT_IP6_PREFIX_STRING_SIZE];
		    otIp6PrefixToString(&config.mPrefix, string, sizeof(string));
		    auto prefix = std::string(string);
    
		    // Remove the prefix length (/32, /64, /96 etc) if present
		    if(prefix.find('/') != std::string::npos)
		    {
			    prefix.erase(prefix.size() - 3);
		    }
    
		    // Remove the last `0:0` if `::` is not found as we are creating a prefix
		    if(prefix.find("::") == std::string::npos)
		    {
			    prefix.erase(prefix.size() - 3);
		    }
    
		    LOG_INF("Found NAT64 prefix: %s", prefix.c_str());
    
		    // Save the prefix in settings and break (@todo save multiple nat64 prefixes)
		    settings::device::set_nat64_prefix(prefix);
    
		    return result;
	    }
	    
            result = otNetDataGetNextRoute(ot_instance, &iterator, &config);
    }
    
    return result;
}
```

**Comment by jwhui** [comment]
> How / when does the device get the external routes?

Border Routers are responsible for publishing the NAT64 prefix / external route to the Leader. The Leader is then responsible for including it in the Network Data and distributing to all Thread devices.

The NAT64 prefix / external route is encoded as part of the Network Data. As part of attaching to the network, the device retrieves the current Network Data.

> Is there a function for the device to request them manually?

Unfortunately, I'm not aware of a API specifically for retrieving a NAT64 prefix / external route. We could consider adding this as a convenience function.

> Do you see any glaring issues with the function below?

The code appears OK at first glance.

> Could this be an issue with the device joining via a leader and not the actual border router (ESP32 based)?

Which specific device the node connects through should not matter. The Network Data should be distributed through all devices.

One option for debugging is to log / monitor changes to the complete Network Data to see if the NAT64 prefix is actually present.

**Comment by motters** [comment]
Hi,

Thanks for the above, it confirms our assumptions. 

Apologies for the very late reply! I've been off work the past week.  We have however done some more testing on this issue the past few days.

It appears the issue only exists when there is more than one device on the network (a leader + router).

We're going to try and reduce the amount of time we have the "openthread_api_mutex_lock(openthread_get_default_context());" locked and see what happens.

I think have a function to manually request the external routes would be super useful.

I'll get back to you shortly after more testing.

Thanks again!

Sam

**Comment by motters** [comment]
So we've been testing this for awhile, trying different things.

It seems that the joining process works as can be seen below:
<img width="1078" alt="image" src="https://github.com/openthread/openthread/assets/2155687/58920574-51d7-4eb0-8f51-fe6a568437c7">

However after joining when the device tries to become a child we get the following:

<img width="1136" alt="image" src="https://github.com/openthread/openthread/assets/2155687/cc3758a9-a6ee-4624-a023-569e84ae3472">

This then repeats with each attempt incrementing,
<img width="445" alt="image" src="https://github.com/openthread/openthread/assets/2155687/e635e635-f59b-4072-b590-329fe1e569fc">
<img width="557" alt="image" src="https://github.com/openthread/openthread/assets/2155687/9ef4c698-47d3-49eb-8c1e-7bb1464a2121">
...

Any advise on what this could be?

**Comment by motters** [comment]
After more debugging it seems the timeout specified in OpenThread for kChildIdResponseTimeout is too short for our device. Tbh our device is doing alot of sampling + dsp.

This issue is only apparent when the new device joining the network does so via an existing device that was elevated to a leader.

From sending the child id request to receiving a response(for our firmware/nrf52840) seems to be taking around ~2.8 seconds:
![Screenshot at Jan 12 15-06-15](https://github.com/openthread/openthread/assets/2155687/9f4f8940-ad7a-4c58-b719-865c08cdc58f)

We have changed the following in mle.hpp which may have solved the issue:
<img width="911" alt="image" src="https://github.com/openthread/openthread/assets/2155687/b89917db-3497-4714-906a-58264362aa49">

We now have a mesh with three devices. I'm going to test with more devices now.

@jwhui  Could i ask why kChildIdResponseTimeout is set to 1250? It seems a little short for a TX/RX event across uCs. Is there an issue with raising this to say 5000?

**Reply by jwhui** [reply, marked as answer]
One of the design goals for Thread is to support quick response times and we expect 1.25 seconds should be more than enough time to turn around a response. I would look into ways of scheduling other operations so that Thread can maintain its responsiveness.

**Comment by motters** [comment]
Seems like page 4-52 of the OT spec states that "MLE_PARENT_REQ_SCANMASK_RE_TIMEOUT" -> "Time an attaching node MUST wait for a Parent Response ..." is 1.25 seconds.

I'm assuming "MUST" means exact and not minimum. So maybe this can't be changed?

**Reply by jwhui** [reply]
Yes, response time requirement is a maximum so that other devices know how long to wait.
